ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs

```diff
@@ -2318,13 +2514,114 @@ prop_inbound_governor_valid_transition_order serverAcc (ArbDataFlow dataFlow)
        )
        verifyRemoteTransitionOrder
     . splitRemoteConns
-    $ evsRTT
+    $ remoteTransitionTraceEvents
   where
     sim :: IOSim s ()
     sim = multiNodeSim serverAcc dataFlow
                        (Script (toBearerInfo absBi :| [noAttenuation]))
                        maxAcceptedConnectionsLimit l
 
+-- | Property wrapping `multinodeExperiment`.
+--
+-- Note: this test validates inbound governor counters.
+--
+prop_inbound_governor_counters :: Int
+                               -> ArbDataFlow
+                               -> MultiNodeScript Int TestAddr
+                               -> Property
+prop_inbound_governor_counters serverAcc (ArbDataFlow dataFlow)
+                               script@(MultiNodeScript l) =
+  let trace = runSimTrace sim
+
+      inboundGovernorEvents :: Trace (SimResult ())
+                        (InboundGovernorTrace
+                          SimAddr)
+      inboundGovernorEvents = traceWithNameTraceEvents trace
+
+      upperBound = multiNodeScriptToCounters l
+
+  in tabulate "ConnectionEvents" (map showConnectionEvents l)
+    . counterexample (ppScript script)
+    . counterexample (Trace.ppTrace show show inboundGovernorEvents)
+    . getAllProperty
+    . bifoldMap
+       ( \ case
+           MainReturn {} -> mempty
+           v             -> AllProperty
+                            $ counterexample (show v) (property False)
+       )
+       ( \ trs
+        -> case trs of
+          TrInboundGovernorCounters igc ->
+            AllProperty
+              $ counterexample
+                  ("Upper bound is: " ++ show upperBound
+                  ++ "\n But got: " ++ show igc)
+                  (property $ igc <= upperBound)
+          _                               ->
+            mempty
+       )
+    $ inboundGovernorEvents
+  where
+    bundleToCounters :: Bundle [Int] -> InboundGovernorCounters
+    bundleToCounters (Bundle hot warm _) =
+      let warmRemote = bool 1 0 (null warm)
+          hotRemote  = bool 1 0 (null hot)
```

Comment:
It's worth to mention that this is an estimate if there is no attenuation.

An idea: would this test pass if we know that there's no attenuation and we changed the upper bound into exact equality?

Reply:
Good question, the answer will be no, because there's a lot of action going on in the state machine and the main reason why it is hard to test the counters with exact equality is because we can not predict when a given connection is going to transition, because that logic is hidden inside the `ConnectionManager`.

Reply:
We could do that if we had better scheduling of events, but this is outside of this pr.